pkg/archive/zip/reader.go

```diff
@@ -0,0 +1,763 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Copyright (c) 2022 Palantir Technologies. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package zip
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"hash/crc32"
+	"io"
+	"io/fs"
+	"os"
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	ErrFormat    = errors.New("zip: not a valid zip file")
+	ErrAlgorithm = errors.New("zip: unsupported compression algorithm")
+	ErrChecksum  = errors.New("zip: checksum error")
+)
+
+// A Reader serves content from a ZIP archive.
+type Reader struct {
+	r             io.ReaderAt
+	Comment       string
+	decompressors map[uint16]Decompressor
+
+	// fileList is a list of files sorted by ename,
+	// for use by the Open method.
+	fileListOnce sync.Once
+	fileList     []fileListEntry
+}
+
+// A ReadCloser is a Reader that must be closed when no longer needed.
+type ReadCloser struct {
+	f *os.File
+	Reader
+}
+
+// A File is a single file in a ZIP archive.
+// The file information is in the embedded FileHeader.
+// The file content can be accessed by calling Open.
+type File struct {
+	FileHeader
+	zip          *Reader
+	zipr         io.ReaderAt
+	headerOffset int64
+	zip64        bool  // zip64 extended information extra field presence
+	descErr      error // error reading the data descriptor during init
+}
+
+func (z *Reader) walk(r io.ReaderAt, size int64, handleFile WalkFn) error {
+	end, err := readDirectoryEnd(r, size)
+	if err != nil {
+		return err
+	}
+	z.r = r
+	z.Comment = end.comment
+	rs := io.NewSectionReader(r, 0, size)
+	if _, err = rs.Seek(int64(end.directoryOffset), io.SeekStart); err != nil {
+		return err
+	}
+	buf := bufio.NewReader(rs)
+
+	// The count of files inside a zip is truncated to fit in a uint16.
+	// Gloss over this by reading headers until we encounter
+	// a bad one, and then only report an ErrFormat or UnexpectedEOF if
+	// the handleFile count modulo 65536 is incorrect.
+	var numFiles uint64
+	for {
+		f := &File{zip: z, zipr: r}
+		err = readDirectoryHeader(f, buf)
+		if err == ErrFormat || err == io.ErrUnexpectedEOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		numFiles++
+		if proceed, err := handleFile(f); err != nil || !proceed {
+			return err
+		}
+	}
+	if numFiles != end.directoryRecords { // only compare 16 bits here
```

Comment:
As per the original code and the comment, I believe end.directoryRecords is truncated and so this will fail if there are more than the max value of a uint16 files. Restoring the uint16() conversions here should fix this.

Reply:
I believe the test `TestOver65kFiles` covers this case.
Max uint16 is `65535`, number of files used in that test case is `65578`.
I've added a counter to check that the `65578` is the number of time the handler is called in this test.
Fancy double checking to see if there's something I'm missing here?

Reply:
I've tried going through the file history, and can't seem to find much information to support why this was being done - the actual end.directoryRecords is a uint64 so could have more than 16 bits in. I would assume this was done to support zip files produced by certain software which presumably used only 16 bits despite the file format allowing for more. Given this is checking for an error case I think it's better to go with the original code, given the uncertainty around what might break if we change this.

Reply:
Okay, here's my best attempt at explaining why the standard library is truncating the values:
- There are two zip file formats, zip and zip64. The former only supports a limited number of files (2^16 worth) while the latter is unlimited.
- However the file count in the older zip format is just a counter. You can write as many records as you like in to the zip and central directory as you want, exceeding the 2^16 maximum count. In this case the count cannot be correct for the number of files actually included in the archive.
- You could declare this to be an error, and cap the older zip format at 2^16 files. There is presumably zip software out there that is happy to produce archives in violation of this.
- Therefore, in order to be as accepting as different zip files as possible, the comparison here is truncated to just 16 bits in order to support archives where more than 2^16 files have been stuffed in to them without using the zip64 format.